heartbeat/autodiscover/builder/hints/heartbeats.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	if monitorConfig != nil {
+		configs := []*common.Config{}
+		for _, cfg := range monitorConfig {
+			if config, err := common.NewConfigFrom(cfg); err == nil {
+				configs = append(configs, config)
+			}
+		}
+		l.logger.Debugf("generated config %+v", configs)
+		// Apply information in event to the template to generate the final config
+		return template.ApplyConfigTemplate(event, configs)
+	}
+
+	tempCfg := common.MapStr{}
+	monitors := l.getMonitors(hints)
+
+	var configs []*common.Config
+	for _, monitor := range monitors {
+		// Every monitor must have a type associated with it
+		if _, ok := monitor[montype]; !ok {
+			continue
+		}
+
+		// If a monitor doesn't have a schedule associated with it then default it.
+		if _, ok := monitor[schedule]; !ok {
+			monitor[schedule] = l.config.DefaultSchedule
+		}
+
+		if procs := l.getProcessors(monitor); len(procs) != 0 {
+			monitor[processors] = procs
+		}
+
+		h, ok := l.getHostsWithPort(monitor, port)
+		if !ok {
+			continue
+		}
+		monitor[hosts] = h
+
+		config, err := common.NewConfigFrom(monitor)
+		if err != nil {
+			l.logger.Debugf("unable to create config from MapStr %+v", tempCfg)
+			return []*common.Config{}
+		}
+		l.logger.Debugf("hints.builder", "generated config %+v", config)
+		configs = append(configs, config)
+	}
+
+	// Apply information in event to the template to generate the final config
+	return template.ApplyConfigTemplate(event, configs)
+}
+
+func (l *heartBeat) getType(hints common.MapStr) common.MapStr {
+	return builder.GetHintMapStr(hints, l.config.Key, montype)
+}
+
+func (l *heartBeat) getSchedule(hints common.MapStr) []string {
+	return builder.GetHintAsList(hints, l.config.Key, schedule)
+}
+
+func (l *heartBeat) getRawConfigs(hints common.MapStr) []common.MapStr {
+	return builder.GetHintAsConfigs(hints, l.config.Key)
+}
+
+func (l *heartBeat) getMonitors(hints common.MapStr) []common.MapStr {
+	raw := builder.GetHintMapStr(hints, l.config.Key, "")
+	if raw == nil {
+		return nil
+	}
+
+	var words, nums []string
+
+	for key := range raw {
+		if _, err := strconv.Atoi(key); err != nil {
+			words = append(words, key)
+			continue
+		} else {
+			nums = append(nums, key)
+		}
+	}
+
+	sort.Strings(nums)
+
+	var configs []common.MapStr
+	for _, key := range nums {
+		rawCfg, _ := raw[key]
+		if config, ok := rawCfg.(common.MapStr); ok {
+			configs = append(configs, config)
+		}
+	}
+
+	defaultMap := common.MapStr{}
+	for _, word := range words {
+		defaultMap[word] = raw[word]
+	}
+
+	if len(defaultMap) != 0 {
+		configs = append(configs, defaultMap)
+	}
+	return configs
+}
+
+func (l *heartBeat) getProcessors(hints common.MapStr) []common.MapStr {
+	return builder.GetConfigs(hints, "", "processors")
+}
+
+func (l *heartBeat) getHostsWithPort(hints common.MapStr, port int) ([]string, bool) {
+	var result []string
+	thosts := builder.GetHintAsList(hints, "", hosts)
+	// Only pick hosts that have ${data.port} or the port on current event. This will make
```

Comment:
Does this make sense for HTTP where we have a default port of 80/443?

Reply:
yes, pods usually dont necessarily have default 80/443 ports. they can have arbitrary port numbers. defaulting may not make best sense.